File: backend/app/models/classroom.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class ClassroomActivity(Base):
    __tablename__ = "classroom_activities"

    id: Mapped[int] = mapped_column(primary_key=True)
    teaching_class_id: Mapped[int | None] = mapped_column(
        ForeignKey("teaching_classes.id", ondelete="SET NULL"), index=True
    )
    course_id: Mapped[int] = mapped_column(ForeignKey("courses.id", ondelete="CASCADE"), index=True)
    chapter_id: Mapped[int] = mapped_column(ForeignKey("chapters.id", ondelete="CASCADE"), index=True)
    created_by: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    question: Mapped[str] = mapped_column(Text, nullable=False)
    minutes: Mapped[int] = mapped_column(Integer, default=8, nullable=False)
    status: Mapped[str] = mapped_column(String(20), default="published", nullable=False)
    created_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), nullable=False)


class ClassroomResponse(Base):
    __tablename__ = "classroom_responses"

    id: Mapped[int] = mapped_column(primary_key=True)
    activity_id: Mapped[int] = mapped_column(ForeignKey("classroom_activities.id", ondelete="CASCADE"), index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    answer: Mapped[str] = mapped_column(Text, nullable=False)
    created_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), nullable=False)


class DiscussionThread(Base):
    __tablename__ = "discussion_threads"

    id: Mapped[int] = mapped_column(primary_key=True)
    teaching_class_id: Mapped[int | None] = mapped_column(ForeignKey("teaching_classes.id", ondelete="CASCADE"), index=True)
    course_id: Mapped[int | None] = mapped_column(ForeignKey("courses.id", ondelete="CASCADE"), index=True)
    chapter_id: Mapped[int | None] = mapped_column(ForeignKey("chapters.id", ondelete="CASCADE"), index=True)
    activity_id: Mapped[int | None] = mapped_column(ForeignKey("classroom_activities.id", ondelete="SET NULL"), index=True)
    author_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[str] = mapped_column(String(20), default="published", nullable=False, index=True)
    is_pinned: Mapped[bool] = mapped_column(default=False, nullable=False, index=True)
    reply_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    last_replied_time: Mapped[datetime | None] = mapped_column(DateTime)
    created_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), nullable=False, index=True)
    updated_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now(), nullable=False)


class DiscussionReply(Base):
    __tablename__ = "discussion_replies"

    id: Mapped[int] = mapped_column(primary_key=True)
    thread_id: Mapped[int] = mapped_column(ForeignKey("discussion_threads.id", ondelete="CASCADE"), index=True)
    author_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    parent_reply_id: Mapped[int | None] = mapped_column(ForeignKey("discussion_replies.id", ondelete="CASCADE"), index=True)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[str] = mapped_column(String(20), default="published", nullable=False, index=True)
    created_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), nullable=False)
    updated_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now(), nullable=False)

File: frontend/src/api/classroom.ts
import { http, type ApiResponse } from './http'

export interface ClassroomActivity {
  id: number
  teaching_class_id: number | null
  course_id: number
  chapter_id: number
  created_by: number
  question: string
  minutes: number
  status: string
  created_time: string
}

export interface ClassroomResponse {
  id: number
  activity_id: number
  user_id: number
  answer: string
  created_time: string
}

export interface DiscussionAuthor { id: number; name: string; role: string }
export interface DiscussionThread {
  id: number; teaching_class_id: number | null; course_id: number | null; chapter_id: number | null; activity_id: number | null
  title: string; content: string; status: string; is_pinned: boolean; reply_count: number
  last_replied_time: string | null; created_time: string; updated_time: string; author: DiscussionAuthor
}
export interface DiscussionReply {
  id: number; thread_id: number; parent_reply_id: number | null; content: string; status: string
  created_time: string; updated_time: string; author: DiscussionAuthor
}

export const classroomApi = {
  list: () => http.get<ApiResponse<ClassroomActivity[]>>('/classroom/activities'),
  publish: (payload: { teaching_class_id: number; course_id: number; chapter_id: number; question: string; minutes: number }) =>
    http.post<ApiResponse<ClassroomActivity>>('/classroom/activities', payload),
  respond: (activityId: number, answer: string) =>
    http.post<ApiResponse<ClassroomResponse>>(`/classroom/activities/${activityId}/responses`, { answer }),
  discussions: (params?: Record<string, unknown>) =>
    http.get<ApiResponse<DiscussionThread[]>>('/classroom/discussions', { params }),
  createDiscussion: (payload: { teaching_class_id?: number; course_id?: number; chapter_id?: number; activity_id?: number; title: string; content: string }) =>
    http.post<ApiResponse<DiscussionThread>>('/classroom/discussions', payload),
  discussionReplies: (threadId: number) =>
    http.get<ApiResponse<DiscussionReply[]>>(`/classroom/discussions/${threadId}/replies`),
  replyDiscussion: (threadId: number, content: string, parentReplyId?: number) =>
    http.post<ApiResponse<DiscussionReply>>(`/classroom/discussions/${threadId}/replies`, { content, parent_reply_id: parentReplyId }),
  updateDiscussion: (threadId: number, payload: { title?: string; content?: string }) =>
    http.patch<ApiResponse<DiscussionThread>>(`/classroom/discussions/${threadId}`, payload),
  deleteDiscussion: (threadId: number) => http.delete<ApiResponse<{ id: number }>>(`/classroom/discussions/${threadId}`),
  updateReply: (replyId: number, content: string) =>
    http.patch<ApiResponse<DiscussionReply>>(`/classroom/discussions/replies/${replyId}`, { content }),
  deleteReply: (replyId: number) => http.delete<ApiResponse<{ id: number }>>(`/classroom/discussions/replies/${replyId}`),
  pinDiscussion: (threadId: number) => http.post<ApiResponse<DiscussionThread>>(`/classroom/discussions/${threadId}/pin`),
  closeDiscussion: (threadId: number) => http.post<ApiResponse<DiscussionThread>>(`/classroom/discussions/${threadId}/close`),
}
